airbyte-integrations/connectors/source-genesys/source_genesys/source.py

```diff
@@ -0,0 +1,308 @@
+    """
+    API Docs: https://developer.genesys.cloud/telephony/telephony-apis
+    """
+
+    primary_key = "id"
+    cursor_field = "dateModified"
+
+    def path(self, **kwargs) -> str:
+        return "telephony/providers/edges/phones"
+
+
+class TelephonyProvidersEdgesSites(GenesysStream):
+    """
+    API Docs: https://developer.genesys.cloud/telephony/telephony-apis
+    """
+
+    primary_key = "id"
+    cursor_field = "dateModified"
+
+    def path(self, **kwargs) -> str:
+        return "telephony/providers/edges/sites"
+
+
+class TelephonyProvidersEdgesTrunks(GenesysStream):
+    """
+    API Docs: https://developer.genesys.cloud/telephony/telephony-apis
+    """
+
+    primary_key = "id"
+    cursor_field = "dateModified"
+
+    def path(self, **kwargs) -> str:
+        return "telephony/providers/edges/trunks"
+
+
+class TelephonyStations(GenesysStream):
+    """
+    API Docs: https://developer.genesys.cloud/telephony/stations-apis
+    """
+
+    primary_key = "id"
+
+    def path(self, **kwargs) -> str:
+        return "stations"
+
+
+class UserUsers(GenesysStream):
+    """
+    API Docs: https://developer.genesys.cloud/useragentman/users/
+    """
+
+    primary_key = "id"
+
+    def path(self, **kwargs) -> str:
+        return "users"
+
+
+class UserGroups(GenesysStream):
+    """
+    API Docs: https://developer.genesys.cloud/useragentman/groups/
+    """
+
+    primary_key = "id"
+    cursor_field = "dateModified"
+
+    def path(self, **kwargs) -> str:
+        return "groups"
+
+
+class SourceGenesys(AbstractSource):
+    @staticmethod
+    def get_connection_response(self, config: Mapping[str, Any]):
+        GENESYS_TENANT_ENDPOINT_MAP: Dict = {
+            "Americas (US East)": "https://login.mypurecloud.com",
+            "Americas (US East 2)": "https://login.use2.us-gov-pure.cloud",
+            "Americas (US West)": "https://login.usw2.pure.cloud",
+            "Americas (Canada)": "https://login.cac1.pure.cloud",
+            "Americas (São Paulo)": "https://login.sae1.pure.cloud",
+            "EMEA (Frankfurt)": "https://login.mypurecloud.de",
+            "EMEA (Dublin)": "https://login.mypurecloud.ie",
+            "EMEA (London)": "https://login.euw2.pure.cloud",
+            "Asia Pacific (Mumbai)": "https://login.aps1.pure.cloud",
+            "Asia Pacific (Seoul)": "https://login.apne2.pure.cloud",
+            "Asia Pacific (Sydney)": "https://login.mypurecloud.com.au",
+        }
+
+        token_refresh_endpoint = GENESYS_TENANT_ENDPOINT_MAP.get(config["tenant_endpoint"])
+        token_refresh_endpoint = token_refresh_endpoint + "/oauth/token"
+        client_id = config["client_id"]
+        client_secret = config["client_secret"]
+        refresh_token = None
+        headers = {"content-type": "application/x-www-form-urlencoded"}
+        data = {"grant_type": "client_credentials", "client_id": client_id, "client_secret": client_secret, "refresh_token": refresh_token}
```

Comment:
I encountered a similar challenge in my development. Question, is it because the `grant_type` is different than the hard-coded one `access_token`? 

https://github.com/airbytehq/airbyte/blob/7fcadf433c9dcecfbc86b69ca9ccc77ad84b32ab/airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/abstract_oauth.py#L49-L54

Reply:
Correct we have to use `grant_type=client_credentials`

Reply:
does it work if you derive [Oauth2Authenticator](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/http/requests_native_auth/oauth.py) and override the hard coded `grant_type`? So that you don't need to re-implement the HTTP request.